streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2472,18 +2472,71 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
     }
 
   /**
-   * Interrupts the stream if it does not produce a value after d duration.
+   * Ends the stream if it does not produce a value after d duration.
    */
   final def timeout(d: Duration): ZStream[R with Clock, E, O] =
     ZStream[R with Clock, E, O] {
       self.process.map { next =>
         next.timeout(d).flatMap {
-          case Some(a) => ZIO.succeedNow(a)
-          case None    => ZIO.interrupt
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.end
         }
       }
     }
 
+  /**
+   * Ends the stream with given error if it does not produce a value after d duration.
+   */
+  final def timeoutError[E1 >: E](e: E1)(d: Duration): ZStream[R with Clock, E1, O] =
+    ZStream[R with Clock, E1, O] {
+      self.process.map { next =>
+        next.timeout(d).flatMap {
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.fail(e)
+        }
+      }
+    }
+
+  /**
+   * Halts the stream with given cause if it does not produce a value after d duration.
+   */
+  final def timeoutErrorCause[E1 >: E](cause: Cause[E1])(d: Duration): ZStream[R with Clock, E1, O] =
+    ZStream[R with Clock, E1, O] {
+      self.process.map { next =>
+        next.timeout(d).flatMap {
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.halt(cause)
+        }
+      }
+    }
+
+  /**
+   * Switches the stream if it does not produce a value after d duration.
+   */
+  final def timeoutTo[R1 <: R, E1 >: E, O2 >: O](
+    d: Duration
+  )(that: ZStream[R1, E1, O2]): ZStream[R1 with Clock, E1, O2] =
+    ZStream[R1 with Clock, E1, O2] {
+      for {
+        currStream   <- Ref.make[ZIO[R1, Option[E1], Chunk[O2]]](Pull.end).toManaged_
+        switchStream <- ZManaged.switchable[R1, Nothing, ZIO[R1, Option[E1], Chunk[O2]]]
+        switched     <- Ref.make(false).toManaged_
+        _            <- switchStream(self.process).flatMap(currStream.set).toManaged_
+        pull = {
+          def go: ZIO[R1 with Clock, Option[E1], Chunk[O2]] =
+            for {
+              switched_ <- switched.get
+              effect    = currStream.get.flatten
+              switch    = switched.set(true) *> switchStream(that.process).flatMap(currStream.set) *> go
+              timeout   = effect.timeout(d).flatMap(_.map(Pull.emit).getOrElse(switch))
```

Comment:
To save a (small) amount of allocations and simplify `go` a bit, these could become `val`s outside of `go`.

Reply:
Done, I had to make those lazy to avoid the `forward reference` compile error.